### DISCUSSION TITLE
Proposal: Improved InternalsVisibleToAttribute

### DISCUSSION THREAD

**Discussion by marek-safar** [dotnet/csharplang#5135] [Language Ideas]
The default InternalsVisibleToAttribute is defined as follows

```c#
[AttributeUsageAttribute(AttributeTargets.Assembly, AllowMultiple = true, Inherited = false)]
public sealed class InternalsVisibleToAttribute : Attribute
```

this restricts the attribute location to assembly level only which means the C# compiler internal->public behaviour cannot be controlled on a more granular level. This leads to problems because all internal types/members are exposed when the attribute is applied and for large assemblies like mscorlib this often leads to conflicts with types from other assemblies.

### Example of the scenario to resolve

mscorlib.dll with IVT to Sample assembly

```C#
namespace System
{
    // This is just a helper class we don't want to expose to anyone
    internal class Utils
    {
    }
}
```

System.dll

```C#
namespace System
{
    public class Utils
    {
    }
}
```

Now if Sample assembly uses both assembly references it will fail to compile with CS0433 error with no easy way to resolve the error.

The proposal is to extend C# compiler analysis of InternalsVisibleTo to also work with following enhanced InternalsVisibleToAttribute which would allow applying the attribute only to necessary types/members instead of the whole assembly.

```c#
[AttributeUsageAttribute(AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Interface | AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Delegate | AttributeTargets.Enum | AttributeTargets.Field | AttributeTargets.Method | AttributeTargets.Property, AllowMultiple = true,     Inherited = false)])]
public sealed class InternalsVisibleToAttribute : Attribute
```

**Comment by DavidArno** [comment]
This would need a CLR change first as the concept of "friend assemblies" is embedded in the runtime, rather than just being a C# feature. So you'd need to raise this as a change to the framework and core CLR's first for it to progress anywhere.

**Comment by tmds** [comment]
C# doesn't allow the `private` modifier on a namespace level `class`.
It could be used to not expose the type through`InternalsVisibleToAttribute`.


```cs
    // This type is not exposed by `InternalsVisibleToAttribute`
    private class Utils
```

**Comment by acaly** [comment]
I think this might fall into the "use analyzer instead of new language feature" category.

**Comment by DavidArno** [comment]
The problem with private classes is twofold: they would need a runtime change and it ought to become the default if no modifier were specified, which would likely be a big breaking change.

The problem with the analyzer approach is that it's too easy to bypass to make it a practical enforcement method.

I've recently started toying with the idea of taking a different approach by using source generators to create public shims of specific classes, eg for a class `Foo` marked with an `{InternalAPI]` attribute, it might creates a `public class Foo_InternalAPIUnderTest : Foo ...` class. This potentially would allow very fine grained control of which internal classes were exposed. There's challenges around handling sealed and static classes for example, but it's definitely a potential solution to this problem.

**Comment by tmds** [comment]
> they would need a runtime change and it ought to become the default if no modifier were specified, which would likely be a big breaking change.

It wouldn't be the default to avoid it being a big breaking change.
